northstar/src/runtime/token.rs

```diff
@@ -96,17 +96,19 @@ fn now() -> time::Duration {
 }
 
 fn calculate_hmac(time: &time::Duration, user: &[u8], target: &[u8], shared: &[u8]) -> Hmac {
-    let mut hasher =
-        HmacSha256::new_from_slice(MAC_KEY.as_slice()).expect("Failed to create SHA-256 hasher");
-    hasher.update(&time.as_millis().to_be_bytes());
-    hasher.update(user);
-    // Sperate user and target
-    hasher.update(&MAC_KEY[0..4]);
-    hasher.update(target);
-    // Sperate target and shared
-    hasher.update(&MAC_KEY[0..4]);
-    hasher.update(shared);
-    hasher.finalize()
+    let mut hmac = HmacSha256::new_from_slice(MAC_KEY.as_slice())
+        .expect("Failed to create SHA-256 HMAC instance");
+    hmac.update(user);
+    let user = hmac.finalize_reset();
+    hmac.update(target);
+    let target = hmac.finalize_reset();
+    hmac.update(shared);
+    let shared = hmac.finalize_reset();
+    hmac.update(&time.as_millis().to_be_bytes());
+    hmac.update(&user.into_bytes());
+    hmac.update(&target.into_bytes());
+    hmac.update(&shared.into_bytes());
```

Comment:
nit: The `into_bytes` call should not be needed.

Reply:
I thought so too but omitting it gives
```
hmac.update(&user);
            ^^^^^ expected slice `[u8]`, found struct `sha2::digest::CtOutput
```

Reply:
`user.as_ref()`?

Reply:
Does not exists. The API of CtOutput is very restricted and does not offer much more than `new` and `into_bytes`:
https://docs.rs/digest/latest/x86_64-apple-darwin/digest/struct.CtOutput.html

The HMAC docs state that `To get underlying array use into_bytes ...`
https://docs.rs/hmac/0.12.1/hmac/

Reply:
Ah ok. Sorry I didn't realise that `user` and `target` is a `CtOutput` not the arg type `AsRef<[u8]>`.